src/lib/reasoners/bitv_rel.ml

```diff
@@ -2229,13 +2269,33 @@ let case_split env uf ~for_model =
         Shostak.Bitv.is_mine @@
         Bitv.extract w bitidx bitidx (Shostak.Bitv.embed r)
       in
-      (* Just always pick zero for now. *)
-      let zero = Shostak.Bitv.is_mine Bitv.[ { bv = Cte Z.zero ; sz = 1 } ] in
+      (* Make sure to pick a value that is compatible with the interval
+         domain.
+
+         This should not happen if cross-domain propagation was run to
+         completion, which it should do, but it is OK if there are corner
+         cases where it didn't (we are just going to be slower), so we
+         want to handle this case.
+
+         Raise an internal error for debugging purposes. *)
+      let value, size =
+        let bit_interval = Intervals.Int.extract int ~ofs:bitidx ~len:1 in
+        match Intervals.Int.value_opt bit_interval with
+        | Some (v, _) ->
+          if Options.get_enable_assertions () then
+            Util.internal_error
+              "High bit allowed by bitlists, but not intervals."
+          else
+            const 1 v, 1
+        | None -> const 1 Z.zero, 2
+      in
```

Comment:
I don't understand why you choose this test. I think that you check if the integer interval is tight enough to constraint the most significant unknown bit of `rr`. Clearly, it is a contradiction if there is such a constraint and you correct your split in this case, right?

Reply:
If the most significant bit unknown bit in the bitlist domain is forced in the interval domain, it should have been forced in the bitlist domain by the cross-propagator (`constrain_bitlist_from_interval`).

If it has somehow *not* been propagated to the proper value from intervals to bitlist, and we pick the wrong value, we wlil loop in the case split. Let's say that the high bit is forced to `1` by the interval domain.

 - We split on `h = 0`; the theory propagates back the equality `h = 0`
 - We find a contradiction
 - Back-tracking: the theory propagates the disequality `h <> 0`
 - This is a disequality (distinct), so it is only added to the interval domain
 - Since the interval domain is not properly cross-propagated to the bitlist domain, we don't see that information and split on `h = 0` again.

Reply:
Ok, I think I can answer my own question. We cannot do better with lower unknown bit.